### PR TITLE
Azure - Add MSI Authentication Support

### LINKED ISSUE #3168: azure - add support for MSI auth
If Custodian is running on a resource with MSI enable, it should be able to authenticated using the sp configured for that resource.

We can use an environment variable to indicate MSI and then use the local MSI endpoint to fetch an AAD token. We will need to be able to request different resource tokens (i.e. storage and arm) from the MSI endpoint.

### PR DESCRIPTION
**Pull request by stefangordon**
closes #3168

Note that MSI is not supported in Azure Linux Functions yet, so this is only useful in scenarios where you are executing Custodian in an Azure VM or App Services instance in some other way.

This supports both system and user identities.

Note that due to the caching behavior of MSI if you run custodian before granting permissions the token will be cached for several hours before reflecting any RBAC changes and you'll need to wait it out...

### CODE REVIEW

File: tools/c7n_azure/c7n_azure/session.py [hunk 1]
```diff
@@ -100,6 +106,17 @@ def _initialize_session(self):
...
             self.log.info("Creating session with Service Principal Authentication")
 
+        elif all(k in os.environ for k in msi_auth_variables):
+            # MSI authentication
+            if constants.ENV_CLIENT_ID in os.environ:
+                self.credentials = MSIAuthentication(
+                    client_id=os.environ[constants.ENV_CLIENT_ID],
+                    resource=self.resource_namespace)
+            else:
+                self.credentials = MSIAuthentication(resource=self.resource_namespace)
```
Comment (logachev) [inline tools/c7n_azure/c7n_azure/session.py]: nit: I think this should also work (just one-liner)
self.credentials = MSIAuthentication(resource=self.resource_namespace, client_id=os.environ.get(constants.ENV_CLIENT_ID, None))

File: tools/c7n_azure/tests/test_session.py [hunk 2]
```diff
@@ -47,6 +48,19 @@ def test_initialize_session_principal(self):
                 self.assertIs(type(s.get_credentials()), ServicePrincipalCredentials)
                 self.assertEqual(s.get_subscription_id(), 'ea42f556-5106-4743-99b0-c129bfa71a47')
 
+    def test_initialize_msi_auth(self):
+        with patch('msrestazure.azure_active_directory.MSIAuthentication.__init__',
+                   autospec=True, return_value=None):
+            with patch.dict(os.environ,
+                            {
+                                constants.ENV_USE_MSI: 'true',
```
Comment (logachev) [inline tools/c7n_azure/tests/test_session.py]: I think we can add a test with client_id being set

File: tools/c7n_azure/c7n_azure/session.py (hunk 1 above)
Comment (stefangordon) [reply tools/c7n_azure/c7n_azure/session.py]: When I did it this way I knew you'd wonder why :)  

The MSIAuthentication class doesn't check for parameters being null, but actually checks for the existence of keys in the argument list so I thought it better to not include it.
https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/azure_active_directory.py#L564

Comment (aluong) [reply tools/c7n_azure/c7n_azure/session.py]: I think you could do:
```
kwargs = {"resource":  self.resource_namespace}
if constants.ENV_CLIENT_ID in os.environ:
  kwargs['client_id'] = os.environ[constants.ENV_CLIENT_ID]

self.credentials = MSIAuthentication(**kwargs)
```

File: tools/c7n_azure/tests/test_session.py [hunk 3]
```diff
@@ -47,6 +48,19 @@ def test_initialize_session_principal(self):
                 self.assertIs(type(s.get_credentials()), ServicePrincipalCredentials)
                 self.assertEqual(s.get_subscription_id(), 'ea42f556-5106-4743-99b0-c129bfa71a47')
```
Comment (aluong) [inline tools/c7n_azure/tests/test_session.py]: You can use DEFAULT_SUBSCRIPTION_ID for the subscription id, it comes from azure_common.

File: tools/c7n_azure/c7n_azure/session.py (hunk 1 above)
Comment (stefangordon) [reply tools/c7n_azure/c7n_azure/session.py]: For sure you could do that :)

**Comment by stefangordon** [comment]
This is ready to merge - just haven't caught it without needing to update branch and wait for a build yet :)

**Pull request merged**